docs/model_developers/plugins.rst

```diff
@@ -14,9 +14,75 @@ plugins with the template engine (Jinja2) to use as filters.
 
 A plugin is a python function, registered with the platform with the :func:`~inmanta.plugins.plugin`
 decorator. This plugin accepts arguments when called from the DSL and can return a value. Both the
-arguments and the return value must by annotated with the allowed types from the orchestration model.
-Type annotations are provided as a string (Python3 style argument annotation). ``any`` is a special
-type that effectively disables type validation.
+arguments and the return value must be annotated with the allowed types from the orchestration model.
+
+To provide this DSL typing information, you can either:
+
+
+Pass the native python type that corresponds to the :term:`DSL` type at hand. e.g. the ``foo`` plugin
+defined below can be used in a model, in a context where the following signature is expected ``string -> int[]``:
+
+
+.. code-block:: python
+    :linenos:
+
+    from inmanta.plugins import plugin
+    from collections.abc import Sequence
+
+
+    @plugin
+    def foo(value: str) -> Sequence[int]:
+        ...
+
+This approach is the recommended way of adding type information to plugins as it allows you to leverage
+type checkers (e.g. mypy) when writing plugin code.
+
+This approach also fully supports the use of ``Union`` types (e.g. ``Union[str, int]`` for an argument
+or a return value, that can be of either type).
+
+
+The table below shows correspondence between types from the Inmanta DSL and their respective python counterpart:
+
+
++------------------+-----------------------------------+
+| Inmanta DSL type | Python type                       |
++==================+===================================+
+| string           | str                               |
++------------------+-----------------------------------+
+| int              | int                               |
++------------------+-----------------------------------+
+| float            | float                             |
++------------------+-----------------------------------+
+| int[]            | collections.abc.Sequence[int]     |
++------------------+-----------------------------------+
+| dict[int]        | collections.abc.Mapping[str, int] |
++------------------+-----------------------------------+
+| string?          | str | None                        |
++------------------+-----------------------------------+
+
+
+
+
+
+Alternatively, the Inmanta :term:`DSL` type annotations can be provided as a string (Python3 style argument annotation).
+The snippet below is equivalent (typing-wise) to the previous one, but this style is discouraged as type-checkers
+won't work out-of-the-box alongside it, and it also doesn't support more elaborate typing constructs such as ``Union``.
```

Comment:
```suggestion
Alternatively, the Inmanta :term:`DSL` type annotations can be provided as a string (Python3 style argument annotation).
```

I would not go too meta (i.e. end users don't need to know all the things we don't support. It only leads to confusion. Unions can simple not be expressed in the dsl....